contracts/epochs/FlowEpoch.cdc

```diff
@@ -109,46 +121,511 @@ pub contract FlowEpochs {
+    /// indexed by epoch number
+    access(contract) var epochMetadata: {UInt64: EpochMetadata}
+
+    access(contract) let QCAdmin: @FlowEpochClusterQC.Admin
+
+    access(contract) let DKGAdmin: @FlowDKG.Admin
+
+    access(contract) let stakingAdmin: @FlowIDTableStaking.Admin
+
+    pub let adminStoragePath: StoragePath
+    pub let heartbeatStoragePath: StoragePath
+
+    /// Resource that can update some of the contract fields
+    pub resource Admin {
+        pub fun updateEpochViews(_ newEpochViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInStakingAuction: "Epoch Views must be greater than staking views"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInDKGPhase: "Epoch Views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInEpoch = newEpochViews
+        }
+
+        pub fun updateAuctionViews(_ newAuctionViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newAuctionViews < FlowEpoch.configurableMetadata.numViewsInEpoch: "Epoch Views must be greater than staking views"
+                newAuctionViews > FlowEpoch.configurableMetadata.numViewsInDKGPhase: "Staking views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInStakingAuction = newAuctionViews
+        }
+
+        pub fun updateDKGPhaseViews(_ newPhaseViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newPhaseViews < FlowEpoch.configurableMetadata.numViewsInEpoch: "Epoch Views must be greater than DKG phase views"
+                newPhaseViews < FlowEpoch.configurableMetadata.numViewsInStakingAuction: "Staking views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInDKGPhase = newPhaseViews
+        }
+
+        pub fun updateNumCollectorClusters(_ newNumClusters: UInt16) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+            }
+
+            FlowEpoch.configurableMetadata.numCollectorClusters = newNumClusters
+        }
+    }
+
+    pub resource Heartbeat {
+
+        /// Function that is called every block to advance the epoch
+        pub fun advanceBlock() {
+
+            let currentBlock = getCurrentBlock()
+            let currentEpochMetadata = FlowEpoch.epochMetadata[FlowEpoch.currentEpochCounter]!
+
+            if FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION {
```

Comment:
This is a method with lots of complexity. At first I wondered why the ifs where not colleapsed into one, why there where two levels, but then I got the nueance of it. 

Have you considered returning and not using else ifs for the other states? switch/case statement on enums would have made this code a lot easier to read.

Reply:
Yeah, I'll probably use `switch` instead. I keep forgetting that exists!

Reply:
Oh Bastian implemented that, I did not know :) Wonderful.